### PR TITLE
PT-476 add cache to promoition evaluation

### PR DESCRIPTION
**Pull request by ksavosteev**
https://virtocommerce.atlassian.net/browse/PT-475

### CODE REVIEW

File: src/VirtoCommerce.MarketingModule.Core/Model/Promotions/PromotionEvaluationContext.cs [hunk 1]
```diff
@@ -80,9 +84,44 @@ public ICollection<string> Coupons
         /// </summary>
         public ProductPromoEntry PromoEntry { get; set; }
 
+        public string GetCacheKey()
+        {
+            return string.Join("|", GetCacheKeyComponents().Select(x => x ?? "null").Select(x => x is ICacheKey cacheKey ? cacheKey.GetCacheKey() : x.ToString()));
```
Comment (Vectorfield4) [inline src/VirtoCommerce.MarketingModule.Core/Model/Promotions/PromotionEvaluationContext.cs:89]: Do we really need so long key?

**Review by Vectorfield4 — APPROVED** [review]

**Comment by Vectorfield4** [comment]
Resolve duplication, please.

File: src/VirtoCommerce.MarketingModule.Data/Services/EvaluationPolicies/BasePromotionPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                cacheEntry.SlidingExpiration = TimeSpan.FromMinutes(1);
+                cacheEntry.AddExpirationToken(PromotionSearchCacheRegion.CreateChangeToken());
+
+                return await EvaluatePromotionCachelessAsync(promoContext);
+            });
+
+            return result;
+        }
+
+        protected abstract Task<PromotionResult> EvaluatePromotionCachelessAsync(PromotionEvaluationContext promoContext);
```
Comment (tatarincev) [inline src/VirtoCommerce.MarketingModule.Data/Services/EvaluationPolicies/BasePromotionPolicy.cs:37]: EvaluatePromotionCachelessAsync -> EvaluatePromotionWithNoCache or EvaluatePromotionWithoutCache (but it is not exactly)

**Review by tatarincev — CHANGES_REQUESTED** [review]
Forgot to click on "Request changes". Ignore this comment please

**Review by tatarincev — APPROVED** [review]

**Comment by artem-dudarev** [comment]
This PR introduces multiple issues with cache keys that were already fixed in the platform:
https://github.com/VirtoCommerce/vc-platform/pull/2129

**Pull request merged**